dhcpv4/server.go

```diff
@@ -0,0 +1,154 @@
+package dhcpv4
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"sync"
+	"time"
+)
+
+/*
+  To use the DHCPv4 server code you have to call NewServer with two arguments:
+  - a handler function, that will be called every time a valid DHCPv4 packet is
+    received, and
+  - an address to listen on.
+
+  The handler is a function that takes as input a packet connection, that can be
+  used to reply to the client; a peer address, that identifies the client sending
+  the request, and the DHCPv4 packet itself. Just implement your custom logic in
+  the handler.
+
+  The address to listen on is used to know IP address, port and optionally the
+  scope to create and UDP socket to listen on for DHCPv4 traffic.
+
+  Example program:
+
+
+package main
+
+import (
+	"log"
+	"net"
+
+	"github.com/insomniacslk/dhcp/dhcpv4"
+)
+
+func handler(conn net.PacketConn, peer net.Addr, m dhcpv4.DHCPv4) {
+	// this function will just print the received DHCPv4 message, without replying
+	log.Print(m.Summary())
+}
+
+func main() {
+	laddr := net.UDPAddr{
+		IP:   net.ParseIP("127.0.0.1"),
+		Port: 67,
+	}
+	server := dhcpv4.NewServer(laddr, handler)
+
+	defer server.Close()
+	if err := server.ActivateAndServe(); err != nil {
+		log.Panic(err)
+	}
+}
+
+*/
+
+// Handler is a type that defines the handler function to be called every time a
+// valid DHCPv4 message is received
+type Handler func(conn net.PacketConn, peer net.Addr, m *DHCPv4)
+
+// Server represents a DHCPv4 server object
+type Server struct {
+	conn       net.PacketConn
+	connMutex  sync.Mutex
+	shouldStop chan bool
+	Handler    Handler
+	localAddr  net.UDPAddr
+}
+
+// LocalAddr returns the local address of the listening socket, or nil if not
+// listening
+func (s *Server) LocalAddr() net.Addr {
+	s.connMutex.Lock()
+	defer s.connMutex.Unlock()
+	if s.conn == nil {
+		return nil
+	}
+	return s.conn.LocalAddr()
+}
+
+// ActivateAndServe starts the DHCPv4 server
+func (s *Server) ActivateAndServe() error {
+	s.connMutex.Lock()
```

Comment:
why not defer the unlock here?

Reply:
because I need it to be unlocked as soon as possible (see after the check on `err`), but also unlock it if something fails (see inside the `err` check)